### DISCUSSION TITLE
A guide to home row mods

### DISCUSSION THREAD

**Discussion by giscus** [precondition/precondition.github.io#26] [Comments]
# A guide to home row mods

What are home row mods? What is a mod-tap? What settings do you need to use? What are the alternatives? And more in this article...

https://precondition.github.io/home-row-mods

**Comment by paulistoan** [comment]
Thank you so much, I found this very useful! Personally, I found your suggested approach of chording with thumb keys -> layers the most suitable for me. Here is my reference KMonad keymap: https://github.com/paulistoan/kmonad-keymaps/blob/main/cags.kbd.

**Comment by kittydoor** [comment]
You are insane for writing such a detailed write up on this! And it is incredibly appreciated!!! I have seen things about this before but always ignored it, because it just seemed weird and too complicated. Now that I actually understand it, and you've explained so many of the tradeoffs / thinking behind it, I'm sold :)

**Comment by constpetrov** [comment]
Thank you so much! Because of the article I did quite a few modifications to my otherwise rusted keyboard configuration, and switched from using vial configurator back to pure qmk fun (my keyboard don't support vial with all needed rules, not enough memory).

This article is indeed an example of how one writes comprehensive yet full and deep guide about not so complex on the surface but surprisingly challenging problem.

**Comment by XeroOl** [comment]
There is a new trick for homerow mods done with combos: `COMBO_MUST_HOLD_MODS`
This define makes it so that if your combo is a modifier, you have to hit both keys within the combo period AND also hold for the tapping term, which is incredibly resistant to misfires.
Here are the docs for more info: https://docs.qmk.fm/#/feature_combo?id=modifier-combos

**Comment by stef-rausch** [comment]
Your post is a gem! It is an excellent introduction with depth -- exactly what I searched for.

I tried `KMonad` and did not succeed on my first try. I will look into it at another stage, for I found `Kanata` (https://github.com/jtroo/kanata/tree/main), which worked out of the box and is well documented for beginners.

**Comment by seonglae** [comment]
Great! This is exact same idea of creating [Intuiter](https://github.com/seonglae/intuiter) project for increasing people's productivity for any keyboard globally

**Comment by knurd75gh** [comment]
Thank you very much for the article. 
I am a kmonad user and have set up the mod, minus the shift keys, as described here.
My problem is, that I can not get it to work with sticky-keys for shift on lsft and rsft.
It only works for the normal keys, not for the home row modded keys.
should it work? Can these two techniques be combined?

**Comment by Erlendms** [comment]
With the help of this magnificent guide and [karabiner.ts](https://github.com/evan-liu/karabiner.ts), I got home row mods to work with Karabiner-Elements. 🙌🏻 

It has the same features as Kmonad, but is much easier to install. (Especially on Mac! I spent literally 10 hours failing to get Kmonad to work...)

I have a [repo with a guide here](https://github.com/Erlendms/karabiner-actions.git), and can make more layouts if anyone wants.

It doesn't have permissive hold and ignore mod tap from QMK - but it has _way_ less lag, and a typing feel i greatly prefer.

**Reply by argenkiwi** [reply]
I put together a [repo](https://github.com/argenkiwi/kenkyo) with the layered keyboard layout I use. I am trying to have the configuration working on different types of keyboard customization software. Keyd and Kanata and practically finished, but I had not found a good way to do home row mods in Karabiner yet. You can see I started implementing it using Goku [here](https://github.com/argenkiwi/kenkyo/blob/main/karabiner/goku/karabiner.edn), but would I be better off trying karabiner.ts? Home row modifiers and layer swapping/pinning are the two main patterns I am missing at the moment.

**Comment by dunyakirkali** [comment]
What a comprehensive write up!

**Comment by Palpatineli** [comment]
I'm using nvim and wsl2 on windows 10 for both coding and prose writing, and did key logging for a few days to find my frequency. Turns out <ctrl> was the most I used by far and the order was actuall ctrl > alt > shift > cmd. Just one data point for your reference.

**Reply by Palpatineli** [reply]
I meant that ctrl was the most used by far.

**Comment by yuanl** [comment]
Thanks for this great guide.

A small issue in [#finding-the-sweet-spot](https://precondition.github.io/home-row-mods#finding-the-sweet-spot). The key alias for `QK_DYNAMIC_TAPPING_TERM_PRINT` should be `DT_PRNT` instead of DT_PRINT. [QMK document here](https://github.com/qmk/qmk_firmware/blob/master/docs/tap_hold.md#dynamic-tapping-term-dynamic-tapping-term).

**Reply by precondition** [reply]
> Thanks for this great guide.
> 
> A small issue in [#finding-the-sweet-spot](https://precondition.github.io/home-row-mods#finding-the-sweet-spot). The key alias for `QK_DYNAMIC_TAPPING_TERM_PRINT` should be `DT_PRNT` instead of DT_PRINT. [QMK document here](https://github.com/qmk/qmk_firmware/blob/master/docs/tap_hold.md#dynamic-tapping-term-dynamic-tapping-term).

that moment when you mistype the name of your own feature... thanks, it is fixed.

**Comment by TurtleMeds** [comment]
Great guide.
Could you add a section for [keyd](https://github.com/rvaiya/keyd)? Here is my config which includes home row mods.
<details><summary>default.conf</summary>

    [ids]
    
    *
    
    [main]
    
    capslock = overload(arrows, esc)
    bookmarks = toggle(gaming)
    
    esc = capslock
    
    # home row mods
    
    a = overloadi(a, overloadt(alt, a, 125), 100)
    s = overloadi(s, overloadt(shift, s, 125), 100)
    d = overloadi(d, overloadt(control, d, 125), 100)
    f = overloadi(f, overloadt(meta, f, 125), 100)
    
    j = overloadi(j, overloadt(meta, j, 125), 100)
    k = overloadi(k, overloadt(control, k, 125), 100)
    l = overloadi(l, overloadt(shift, l, 125), 100)
    ; = overloadi(;, overloadt(alt, ;, 125), 100)
    
    [gaming]
    
    a = a
    s = s
    d = d
    f = f
    
    j = j
    k = k
    l = l
    ; = ;
    
    [arrows]
    
    h = left
    j = down
    k = up
    l = right
</details>

**Reply by argenkiwi** [reply]
I started a discussion in [kanata](https://github.com/jtroo/kanata/discussions/1455#discussion-7766783) on how to skip home row modifiers completely when typing at speed to prevent misfires. It was actually inspired by keyd's `overloadi` action. It would be great if they were added to this post in some way.

**Comment by isaaceliape** [comment]
Hey everyone! 

For those of you who prefer to listen rather than read, I've put together a "Podcast-like" audio version of this article using Google Notebook LM. You can check it out and listen here:

### [A guide to home row mods (Podcast)](https://notebooklm.google.com/notebook/c53676e4-038c-44a1-a4ec-890673f97fd7/audio)

**Comment by venomNewtona** [comment]
Hi, liked your articles, and I thought maybe you can give me advice.

I'm getting more and more upset as it's 2025 but I still don't see anyone trying to reduce the disadvantage of having to switch between two languages, the English and the spoken one. 

It puts me on tilt when I imagine how much smoother the experience of a person who does not has to live through this lang switching pain ...

Why there is no solution yet? Like the option to connect the smaller native language keyboard and stop being pissed by modality trouble.
The idea is very attracting for me, but it appears there is no way to work around few technical issues. Is there any hope visible from your point of view?

**Reply by precondition** [reply]
Hello, thanks for the comment. It is hard to talk about the issue unless you specify what is your “smaller native language”. I personally write in English, French and Japanese while keeping the exact same keyboard layout and I feel satisfied with my setup. 

If you need to write in two languages based on different alphabets, switching keyboard layout is probably the best solution. You could technically create an IME that would let you compose foreign characters by using roman letters (e.g. type <kbd>z</kbd><kbd>h</kbd><kbd>Space</kbd>, get “ж”) but that solution only really makes sense for non-alphabetic languages such as Mandarin or Japanese.

 Regardless, isn't this issue unrelated to home row mods?

**Reply by venomNewtona** [reply]
my question is really an off-topic, I'm just asking it to every keyboard enthusiast met, in hope eventually someone directs me to the right place. Thanks for replying!

**Comment by mslonik** [comment]
Thank you for this extended and comprehensive article 👍 🧡.

For those of us who stick to Microsoft Windows, there is another interesting alternative, based on [AutoHotkey](https://www.autohotkey.com/) in order to mimic QMK: [QMK.ahk](https://github.com/thehafenator/QMK.ahk.) Already tried for home rows mods by me.

Happy typing, everyone!

**Reply by mslonik** [reply]
Additionally, there is a thread on AutoHotkey forum dedicated to QMK.ahk [here](https://www.autohotkey.com/boards/viewtopic.php?f=83&t=139344&hilit=QMK).
